lib/AFP/OsString/Internal/Types.hs

```diff
@@ -0,0 +1,176 @@
+#elif MIN_VERSION_template_haskell(2,16,0)
+  liftTyped = TH.unsafeTExpCoerce . TH.lift
+#endif
+
+instance Show WindowsString where
+  show (WS bs) = ('\"': decodeUtf16LE bs) <> "\""
+
+instance Show PosixString where
+  show (PS bs) = ('\"': decodeUtf8 bs) <> "\""
+
+instance IsString WindowsString where 
+    fromString = WS . encodeUtf16LE
+
+instance IsString PosixString where 
+    fromString = PS . encodeUtf8
+
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+type PlatformString = WindowsString
+#else
+type PlatformString = PosixString
+#endif
+
+newtype WindowsWord = WW Word16
+  deriving (Eq, Ord, Show)
+newtype PosixWord   = PW Word8
+  deriving (Eq, Ord, Show)
+
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+type PlatformWord = WindowsWord
+#else
+type PlatformWord = PosixWord
+#endif
+
+
+-- | Newtype representing operating system specific strings.
+--
+-- Internally this is either 'WindowsFilePath' or 'PosixFilePath',
+-- depending on the platform. Both use unpinned
+-- 'ShortByteString' for efficiency and correctness.
+--
+-- The constructor is only exported via "AFP.OsString.Internal.Types", since
+-- dealing with the internals isn't generally recommended, but supported
+-- in case you need to write platform specific code, such as the implementation
+-- of 'fromAbstractFilePath'.
+newtype OsString = OsString PlatformString
+  deriving Show
+
+-- | Byte equality of the internal representation.
+instance Eq OsString where
+  (OsString a) == (OsString b) = a == b
+
+-- | Byte ordering of the internal representation.
+instance Ord OsString where
+  compare (OsString a) (OsString b) = compare a b
+
+-- | Encodes as UTF16 on windows and UTF8 on unix.
+instance IsString OsString where 
+    fromString = toOsString
+
+-- | \"String-Concatenation\" for 'AbstractFilePath'. This is __not__ the same
+-- as '(</>)'.
+instance Monoid OsString where 
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+    mempty      = OsString (WS BS.empty)
+    mappend (OsString (WS a)) (OsString (WS b))
+      = OsString (WS (mappend a b))
+#else
+    mempty      = OsString (PS BS.empty)
+    mappend (OsString (PS a)) (OsString (PS b))
+      = OsString (PS (mappend a b))
+#endif
+#if MIN_VERSION_base(4,11,0)
+instance Semigroup OsString where 
+    (<>) = mappend
+#endif
+
+
+instance Lift OsString where
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+  lift (OsString (WS bs))
+    = [| OsString (WS (BS.pack $(lift $ BS.unpack bs))) :: OsString |]
+#else
+  lift (OsString (PS bs))
+    = [| OsString (PS (BS.pack $(lift $ BS.unpack bs))) :: OsString |]
+#endif
+#if MIN_VERSION_template_haskell(2,17,0)
+  liftTyped = TH.unsafeCodeCoerce . TH.lift
+#elif MIN_VERSION_template_haskell(2,16,0)
+  liftTyped = TH.unsafeTExpCoerce . TH.lift
+#endif
+
+
+-- | Newtype representing operating system Word with respect to
+-- the encoding. On Windows, this is 'Word16', on POSIX 'Word8'.
+newtype OsWord = OsWord PlatformWord
```

Comment:
I'm not sure this is totally clear in use case. There are some operations such as:

```hs
isExtSeparator :: Char -> Bool
```

That have to be translated for Windows UTF16 bytestrings.